### ISSUE TITLE
.NET Standard support?

### ISSUE THREAD

**Issue by ctolkien** [intuit/QuickBooks-V3-DotNET-SDK#9]
Has this been evaluated?

**Comment by nimisha84** [comment]
I'm not sure if I understand your question well. Can you elaborate more?

**Comment by ctolkien** [comment]
Currently the Nuget release is targetting .NET Framework 4.0 - aka "Full framework".

To allow us to use this code on other versions of .NET (like .NET Core), it needs to be targetted (and compatible) with .NET standard:
https://docs.microsoft.com/en-us/dotnet/standard/net-standard

**Comment by nimisha84** [comment]
Thanks for the feedback. We can look into this for a future release. However we did check the portability of the current .Net SDK(Framework 4.0) to .Net Core and found that many of the lib are not supported. So, not sure if moving to .Net Standard will really help your use case.
Let me know what you think.

**Comment by ctolkien** [comment]
>So, not sure if moving to .Net Standard will really help your use case.

Well, we're running against .NET Core. This is the only library we're lacking compatibility with. I had a quick look at pulling this codebase down to see how involved it would be, however the master branch does not build.

**Comment by nimisha84** [comment]
Yes, we did do a Portability check for .Net core migration for our project and found that many of the lib of the .Net SDK are not compatible for .Net Core. Some of them did not even have an alternative.
So, this project will not build in .Net Core unfortunately unless most parts of it are rewritten for .Net Core. Currently this work is not prioritized yet for the current SDK. We definitely would be taking caring of this feedback for the next version of the SDK is not started yet.

**Comment by jsgoupil** [comment]
@nimisha84 any plans to revisit this issue? Based on this announcement: https://github.com/aspnet/Announcements/issues/324

QuickBooks would be able to reach further customers and keep up to date with the technology if they could support a more common denominator. 
Your software is not just QuickBooks Desktop, platform dependant, but Online reaches a lot more people, and definitely more than just windows.

**Comment by jsgoupil** [comment]
@nimisha84 Ping?

**Comment by nimisha84** [comment]
@jsgoupil I just returned back from parental leave and saw your question.  Good that you brought this up because we were actually trying to move over our SDK to .Net Standard. Given .Net core would be only supported framework, we will have to reconsider our approach and probably invest on building a new SDK for .Net core only. At this time, I cannot comment on what is the near future plan but you did raise a valid point and we will work on this.

**Comment by jsgoupil** [comment]
@nimisha84 Thank you.
Can you please re-open this bug, so we know it is being tracked.

**Comment by nimisha84** [comment]
Our internal feature req num is IDG-2638 where we are tracking this request. I will reopen this too.
Thanks!

**Comment by luismts** [comment]
This is a bit strange. .net core/standard already have time in the market and it would be normal for all platforms to support it since Microsoft is focusing all its resources on the .net core and .net standard lib. 

This is a bit strange, now we will have to do everything manually. What strikes me most is that they are doing the SDK since 2017, according to what I have seen, and we are already in 2019.

**Comment by nimisha84** [comment]
Hi @luismts Thanks for feedback. The .Net SDK was released in 2010 but was open sourced later on here.
When we open sourced the SDK, we did do a portability analysis and found that only 65% of the libraries we needed were supported in .Net Standard.
So, the migration was not done until all lib were supported in .Net Standard. 
After that due to change in priorities geared towards a V4 SDK support for new V4 API, this work was not taken up. Since, there is still time in V4 API release, we have already started migration on Standard branch as we see not 95% of lib we need are supported in .Net Standard.
 Hope this info helps you.

**Comment by nimisha84** [comment]
@here We are working on releasing the beta version of our SDK DLLS for .net standard which you can use with .Net Core 2.1 projects.
Please let me know if you are interested in testing those out and giving us your feedback if you see any issues.

Thanks,
Nimisha

**Comment by nimisha84** [comment]
Please reply with your email ids, in case you are interested. I can send the zipped SDK dlls and steps to download any additional dependencies from Nuget. Later, in a few weeks when we release it, we will be releasing the NUget package with required dependencies.

**Comment by jsgoupil** [comment]
contact@jsgoupil.com
I will give it a try, but I am really busy right now, so I might not be able to give feedback quick enough. I appreciate the work!

**Comment by IlyaErmakov** [comment]
@nimisha84 , my email is ilya.ermakov@servicetitan.com. Thanks.

**Comment by nimisha84** [comment]
Sorry, thought of creating a NUget release instead. You can search for this package in pre-release on Nuget-
IppDotNetSdkForQuickBooksApiV3_NetStandard2_0 7.5.1-beta (for FULL v3 SDK), IntuitOAuth2PlatformClient_NetStandard2_0 7.5.1-beta (for just Oauth2 SDK).

Use this appsettings.json as a reference for your .Net Core projects-
{
"Logger": {
"RequestLog": {
"EnableLogs": "false",
"LogDirectory": ""
}
},
//not supported added for backward compatibility
"CustomLogger": {
"RequestLog": {
"Enable": "false",
"Name": "",
"Type": ""
}
},
"Security": {
"Mode": {
"Oauth": {
"Enable": "false",
"AccessToken": ""
},
//not supported added for backward compatibility
"Custom": {
"Enable": "false",
"Name": "",
"Type": "",
"Params": ""
}
}
},
"Message": {
"Request": {
"CompressionFormat": "None",
"SerializationFormat": "Json"
},
"Response": {
"CompressionFormat": "None",
"SerializationFormat": "Json"
}
},
"Service": {
"BaseUrl": {
"Qbo": "",
"Ips": "", //not supported added for backward compatibility
"OAuthAccessTokenUrl": "", //not supported added for backward compatibility
"UserNameAuthentication": "" //not supported added for backward compatibility
},
"MinorVersion": {
"Qbo": "37"
}
},
"WebhooksService": {
"VerifierToken": {
"Value": ""
}
},
"Retry": {
"Mode": {
"LinearRetry": {
"Enable": "false",
"RetryCount": "",
"RetryInterval": ""
},
"IncrementalRetry": {
"Enable": "false",
"RetryCount": "",
"InitialInterval": "",
"Increment": ""
},
"ExponentialRetry": {
"Enable": "true",
"RetryCount": "",
"MinBackoff": "",
"MaxBackoff": "",
"DeltaBackoff": ""
}
}
}
}

If you need 4.7.2 dlls for this .Net Standard V3 SDK, then I can share those in a zip file.

**Comment by vchirikov** [comment]
LinqExtender is missing, do you have `Intuit.Ipp.LinqExtender` library in open source? 
p.s. why you copy https://github.com/mehfuzh/LinqExtender/ code without copyright?

**Comment by nimisha84** [comment]
LINQ extender was deprecated long back. We do not use it or support it. It was added when the SDK was first built but not open sourced and it was not having the right copyright credit, it was probably due to lack of knowledge by the person who built it. Thus, we have removed it even from Standard DLLs. We also have removed deprecated Platformclient and Oauth1 support as that is going to be deprecated too. Let me know if see any other issues

**Comment by nimisha84** [comment]
LINQ extender was deprecated long back. We do not use it or support it. It was added when the SDK was first built but not open sourced and it was not having the right copyright credit, it was probably due to lack of knowledge by the person who built it. Thus, we have removed it even from Standard DLLs. We also have removed deprecated Platformclient and Oauth1 support as that is going to be deprecated too. Let me know if see any other issues
